### PR TITLE
Improved the performance of RecallAtFixedPrecision for large batch sizes

### PR DESCRIPTION
**Pull request by GlavitsBalazs**
## What does this PR do?

Addresses #2041.

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2042.org.readthedocs.build/en/2042/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@GlavitsBalazs nice, do you have a benchmark just to give an idea about the speedup? Also pls add entry in changelog

**Comment by GlavitsBalazs** [comment]
> @GlavitsBalazs nice, do you have a benchmark just to give an idea about the speedup? Also pls add entry in changelog

Ok. I've added the changelog entry.

My [issue submission](https://github.com/Lightning-AI/torchmetrics/issues/2041) contains some basic measurements illustrating the speedup. Please let me know if these aren't enough. I'll create more benchmarks if necessary.

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
> > @GlavitsBalazs nice, do you have a benchmark just to give an idea about the speedup? Also pls add entry in changelog
> 
> Ok. I've added the changelog entry.
> 
> My [issue submission](https://github.com/Lightning-AI/torchmetrics/issues/2041) contains some basic measurements illustrating the speedup. Please let me know if these aren't enough. I'll create more benchmarks if necessary.

@GlavitsBalazs I think the tests you done are fine, I am reposting here for future reference:

| Batch Size | Before PR | After PR
|------------|---------------|-----------|
| 100        | ~9            |~8 |
| 1000       | ~20           |~9 |      
| 2000       | ~30           |~10 |
| 5000       | ~70           |~14 |
| 10000      | ~124          |~19 |
| 100000     |  ~1200        | ~110 |

File: src/torchmetrics/functional/classification/recall_fixed_precision.py [hunk 1]
```diff
@@ -37,21 +37,37 @@
...
+
+    """
+    idx: Optional[Tensor] = None
+    for k in range(x.shape[1]):
+        col: Tensor = x[idx, k] if idx is not None else x[:, k]
+        z = torch.where(col == col.max())[0]
+        idx = z if idx is None else idx[z]
+        if len(idx) < 2:
+            break
+    return idx
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/classification/recall_fixed_precision.py:53]:
```suggestion
    if idx is None:
        raise ValueError("Failed to extract index")
    return idx
```
to fix mypy error

**Comment by GlavitsBalazs** [comment]
Closes #2041

**Pull request merged**